runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
 	Trc_MM_SchedulingDelegate_calculateEdenSize_Exit(env->getLanguageVMThread(), (_edenRegionCount * regionSize));
 }
 
-UDATA
+void
+MM_SchedulingDelegate::moveTowardRecommendedEden(MM_EnvironmentVLHGC *env, double edenChangeSpeed) 
+{
+	Assert_MM_true(edenChangeSpeed <= 1 && edenChangeSpeed >= 0);
+
+	if ((0 == _historicalPartialGCTime) || (0 == _averagePgcInterval)) {
+		/* Until we have collected any information about PGC time, we don't have the data we need to make informed decision about eden size  */
+		return;
+	}
+
+	uintptr_t recommendedEdenSizeBytes = calculateRecommendedEdenSize(env);
+
+	uintptr_t currentIdealEdenBytes = getIdealEdenSizeInBytes(env);
+	uintptr_t currentIdealEdenRegions = _idealEdenRegionCount;
+
+	/* 
+	 * The closer edenChangeSpeed is to 1, the larger the move towards recommendedEdenSizeBytes will be. 
+	 * 1 implies that eden should move all the way towards recommendedEdenSizeBytes.
+	 */
+	intptr_t edenChange = (intptr_t)recommendedEdenSizeBytes - currentIdealEdenBytes;
+	intptr_t targetEdenChange = (intptr_t)(edenChange * edenChangeSpeed);
+	uintptr_t targetEdenBytes = currentIdealEdenBytes + targetEdenChange;
+	uintptr_t targetEdenRegions = targetEdenBytes / _regionManager->getRegionSize();
+
+	_edenSizeFactor = (intptr_t)targetEdenRegions - currentIdealEdenRegions;
+}
+
+void
+MM_SchedulingDelegate::checkEdenSizeAfterPgc(MM_EnvironmentVLHGC *env, bool globalSweepHappened)
+{
+	if (!_extensions->statupPhaseFinished) {
+		/* Don't change eden size during startup phase - keep it at default */
+		return;
+	}
+
+	if (_currentlyPerformingGMP && !globalSweepHappened) {
+		/* 
+		 * Don't change eden size while GMP cycle is running - 
+		 * Unless a global sweep just happened, in which case we allow eden to change on first PGC after GMP 
+		 */
+		return;
+	}
+
+	if (heapIsFullyExpanded(env)) {
+		/* 
+		 * When heap is full, try to get eden to such a size to minimize overhead, while keeping into account the limits imposed by
+		 * the heap size, and how much free space is available, since "blindly" increasing eden is no longer an option
+		 */
+		if (globalSweepHappened) {
+			/* 
+			 * Take a more aggreessive step towards ideal eden. 
+			 * At this point we have the most accuate information about liveness in the heap, so we make the most informed decision
+			 */
+			moveTowardRecommendedEden(env, 0.5);
+			resetPgcTimeStatistics(env);
+		} else if ((_pgcCountSinceGMPEnd & (consecutivePGCToChangeEden - 1)) == 0) {
+			/* Every consecutivePGCToChangeEden number of pgc's, re-evaluate eden size, and move towards it */
+			moveTowardRecommendedEden(env, 0.25);
+		}
+	} else if (0 == _pgcCountSinceGMPEnd % 3) {
```

Comment:
I would be comfortable trying resize here on every PGC. It may be relevant during rampup phase. On the other side (we we have wide hysteresis bounds to prevent noise behavior at a stable state.

Reply:
BTW, scavenger in Gencon tries on each one.

The motivation for limiting attempts in fully expanded heap is to avoid recalculations for de-fragmenation work and GMP scheduling, although we can consider lifting the limitation there as well, at some point.

Reply:
During testing/experimentation, I found that if we do this on every PGC, `_historicalPartialGCTime` doesn't have enough time to settle (as you point out). If PGC time is just slightly above pause time target, it can cause eden to decrease by an extra ~20% (ie, 3 decreases of 10%, rather than just 1 of 10%). 

Since `_historicalPartialGCTime` is historically weighed at 80%, it takes a while for values to settle, so my thought was that 3 pgc's would at least be halfway between old historical average, and new average - which would at least keep eden size from accidentally becoming too big/small too fast without an actual need.

Reply:
Then I'd consider reducing weight on old value (0.6 or 0.5) and let it learn new one faster. I'd rather deal with one param (only weight) than 2 (wieght and PGC skip count).
I don't see a problem getting 1 or even 2 extra adjustments due to inertia, as long as they still toward the CPU overhead optimum and not over shooting toward the other extreme. While CPU overhead bounds are rather wide, so overshoot should not occur that easily, I can that occurring for pause threshold (which has only one bound). Still it's a soft threshold, so we should be ok there, but you can play with the waight of pause threshold too to minimized those extra inertial adjustments.